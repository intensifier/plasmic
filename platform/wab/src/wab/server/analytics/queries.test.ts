import {
  AnalyticsQueryConstants,
  getConversionRateQuery,
  getImpressionsQuery,
} from "@/wab/server/analytics/queries";
import { Parser } from "node-sql-parser";

function fillQueryParams(query: string, PARAMS: Record<string, string>) {
  let result = query;
  Object.keys(PARAMS).forEach((param) => {
    const pattern = AnalyticsQueryConstants[param];
    result = result.replace(new RegExp(pattern, "g"), PARAMS[param]);
  });
  return result;
}

const PARAMS = {
  TEAM_ID: `'"gdYGJk5bkBPVEmFndhpoz4"'`,
  PROJECT_ID: `'"kjuFFTSZb8fanzCHT2C1jz"'`,
  TIMEZONE: "'America/Bahia'",
  FROM: "'2022-06-01 00:00:00'",
  TO: "'2022-07-30 23:59:99'",
  COMPONENT_ID: "'provider'",
  SPLIT_ID: "'splitId'",
};

function parseQuery(query: string) {
  const parser = new Parser();
  return parser.parse(query);
}

describe("Analytics queries", () => {
  describe("getImpressionsQuery", () => {
    it("should generate team impressions per day", () => {
      const result = fillQueryParams(
        getImpressionsQuery({
          teamId: PARAMS.TEAM_ID,
          from: PARAMS.FROM,
          to: PARAMS.TO,
          timezone: PARAMS.TIMEZONE,
          period: "day",
        }),
        PARAMS
      );
      expect(result).toMatchSnapshot();
      expect(() => parseQuery(result)).not.toThrow();
    });

    it("should generate team impressions per month", () => {
      const result = fillQueryParams(
        getImpressionsQuery({
          teamId: PARAMS.TEAM_ID,
          from: PARAMS.FROM,
          to: PARAMS.TO,
          timezone: PARAMS.TIMEZONE,
          period: "month",
        }),
        {
          ...PARAMS,
          FROM: "'2022-06-01'",
          TO: "'2022-07-30'",
        }
      );
      expect(result).toMatchSnapshot();
      expect(() => parseQuery(result)).not.toThrow();
    });

    it("should generate project impressions per day sliced", () => {
      const result = fillQueryParams(
        getImpressionsQuery({
          teamId: PARAMS.TEAM_ID,
          from: PARAMS.FROM,
          to: PARAMS.TO,
          timezone: PARAMS.TIMEZONE,
          projectId: PARAMS.PROJECT_ID,
          splitId: PARAMS.SPLIT_ID,
          period: "day",
        }),
        PARAMS
      );
      expect(result).toMatchSnapshot();
      expect(() => parseQuery(result)).not.toThrow();
    });

    it("should generate project impressions per month sliced", () => {
      const result = fillQueryParams(
        getImpressionsQuery({
          teamId: PARAMS.TEAM_ID,
          from: PARAMS.FROM,
          to: PARAMS.TO,
          timezone: PARAMS.TIMEZONE,
          projectId: PARAMS.PROJECT_ID,
          splitId: PARAMS.SPLIT_ID,
          period: "month",
        }),
        {
          ...PARAMS,
          FROM: "'2022-06-01'",
          TO: "'2022-07-30'",
        }
      );
      expect(result).toMatchSnapshot();
      expect(() => parseQuery(result)).not.toThrow();
    });
  });

  describe("getConversionRateQuery", () => {
    it("should generate team conversion rate per day", () => {
      const result = fillQueryParams(
        getConversionRateQuery({
          teamId: PARAMS.TEAM_ID,
          from: PARAMS.FROM,
          to: PARAMS.TO,
          timezone: PARAMS.TIMEZONE,
          period: "day",
        }),
        PARAMS
      );
      expect(result).toMatchSnapshot();
      expect(() => parseQuery(result)).not.toThrow();
    });

    it("should generate team conversion rate per month", () => {
      const result = fillQueryParams(
        getConversionRateQuery({
          teamId: PARAMS.TEAM_ID,
          from: PARAMS.FROM,
          to: PARAMS.TO,
          timezone: PARAMS.TIMEZONE,
          period: "month",
        }),
        {
          ...PARAMS,
          FROM: "'2022-06-01'",
          TO: "'2022-07-30'",
        }
      );
      expect(result).toMatchSnapshot();
      expect(() => parseQuery(result)).not.toThrow();
    });

    it("should generate project conversion rate per day sliced", () => {
      const result = fillQueryParams(
        getConversionRateQuery({
          teamId: PARAMS.TEAM_ID,
          from: PARAMS.FROM,
          to: PARAMS.TO,
          timezone: PARAMS.TIMEZONE,
          projectId: PARAMS.PROJECT_ID,
          splitId: PARAMS.SPLIT_ID,
          period: "day",
        }),
        PARAMS
      );
      expect(result).toMatchSnapshot();
      expect(() => parseQuery(result)).not.toThrow();
    });

    it("should generate project conversion rate per month sliced", () => {
      const result = fillQueryParams(
        getConversionRateQuery({
          teamId: PARAMS.TEAM_ID,
          from: PARAMS.FROM,
          to: PARAMS.TO,
          timezone: PARAMS.TIMEZONE,
          projectId: PARAMS.PROJECT_ID,
          splitId: PARAMS.SPLIT_ID,
          period: "month",
        }),
        {
          ...PARAMS,
          FROM: "'2022-06-01'",
          TO: "'2022-07-30'",
        }
      );
      expect(result).toMatchSnapshot();
      expect(() => parseQuery(result)).not.toThrow();
    });
  });
});
