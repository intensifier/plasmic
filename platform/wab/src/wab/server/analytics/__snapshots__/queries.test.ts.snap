// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytics queries getConversionRateQuery should generate project conversion rate per day sliced 1`] = `"select _moment as time, slice, countIf(maxSteps >= 1) as renders, countIf(maxSteps = 2) as conversions, if (renders > 0, conversions / renders, 0) as conversion_rate from ( select distinct_id, toStartOfDay(timestamp) as _moment, slice, max(steps) as maxSteps from ( select distinct_id, timestamp, slice, if(latest_render < latest_conversion, 2, if(latest_render is not null, 1, 0)) as steps from ( select distinct_id, timestamp, latest_render, slice, min(latest_conversion) over ( partition by ( distinct_id, slice, toStartOfDay(timestamp) ) order by timestamp DESC ) latest_conversion from ( select distinct_id, timestamp, if(event = '$render', timestamp, null) as latest_render, if(event = '$conversion', timestamp, null) as latest_conversion ,JSONExtractString(events.properties, 'splitId') as slice from events where ( ( event = '$render' and has(JSONExtractArrayRaw(events.properties, 'teamIds'), '"gdYGJk5bkBPVEmFndhpoz4"') and has(JSONExtractArrayRaw(events.properties, 'projectIds'), '"kjuFFTSZb8fanzCHT2C1jz"') and JSONExtractString(events.properties, 'rootComponentId') = 'provider' and JSONHas(events.properties, 'splitId') ) or ( event = '$conversion' and has(JSONExtractArrayRaw(events.properties, 'projectIds'), '"kjuFFTSZb8fanzCHT2C1jz"') and JSONHas(events.properties, 'splitId') ) ) and toStartOfDay(events.timestamp, 'America/Bahia') >= '2022-06-01 00:00:00' and toStartOfDay(events.timestamp, 'America/Bahia') <= '2022-07-30 23:59:99' ) ) ) group by distinct_id, toStartOfDay(timestamp) ,slice ) group by time ,slice"`;

exports[`Analytics queries getConversionRateQuery should generate project conversion rate per month sliced 1`] = `"select _moment as time, slice, countIf(maxSteps >= 1) as renders, countIf(maxSteps = 2) as conversions, if (renders > 0, conversions / renders, 0) as conversion_rate from ( select distinct_id, toStartOfMonth(timestamp) as _moment, slice, max(steps) as maxSteps from ( select distinct_id, timestamp, slice, if(latest_render < latest_conversion, 2, if(latest_render is not null, 1, 0)) as steps from ( select distinct_id, timestamp, latest_render, slice, min(latest_conversion) over ( partition by ( distinct_id, slice, toStartOfMonth(timestamp) ) order by timestamp DESC ) latest_conversion from ( select distinct_id, timestamp, if(event = '$render', timestamp, null) as latest_render, if(event = '$conversion', timestamp, null) as latest_conversion ,JSONExtractString(events.properties, 'splitId') as slice from events where ( ( event = '$render' and has(JSONExtractArrayRaw(events.properties, 'teamIds'), '"gdYGJk5bkBPVEmFndhpoz4"') and has(JSONExtractArrayRaw(events.properties, 'projectIds'), '"kjuFFTSZb8fanzCHT2C1jz"') and JSONExtractString(events.properties, 'rootComponentId') = 'provider' and JSONHas(events.properties, 'splitId') ) or ( event = '$conversion' and has(JSONExtractArrayRaw(events.properties, 'projectIds'), '"kjuFFTSZb8fanzCHT2C1jz"') and JSONHas(events.properties, 'splitId') ) ) and toStartOfMonth(events.timestamp, 'America/Bahia') >= '2022-06-01' and toStartOfMonth(events.timestamp, 'America/Bahia') <= '2022-07-30' ) ) ) group by distinct_id, toStartOfMonth(timestamp) ,slice ) group by time ,slice"`;

exports[`Analytics queries getConversionRateQuery should generate team conversion rate per day 1`] = `"select _moment as time, countIf(maxSteps >= 1) as renders, countIf(maxSteps = 2) as conversions, if (renders > 0, conversions / renders, 0) as conversion_rate from ( select distinct_id, toStartOfDay(timestamp) as _moment, max(steps) as maxSteps from ( select distinct_id, timestamp, if(latest_render < latest_conversion, 2, if(latest_render is not null, 1, 0)) as steps from ( select distinct_id, timestamp, latest_render, min(latest_conversion) over ( partition by ( distinct_id, toStartOfDay(timestamp) ) order by timestamp DESC ) latest_conversion from ( select distinct_id, timestamp, if(event = '$render', timestamp, null) as latest_render, if(event = '$conversion', timestamp, null) as latest_conversion from events where ( ( event = '$render' and has(JSONExtractArrayRaw(events.properties, 'teamIds'), '"gdYGJk5bkBPVEmFndhpoz4"') and JSONExtractString(events.properties, 'rootComponentId') = 'provider' ) or ( event = '$conversion' ) ) and toStartOfDay(events.timestamp, 'America/Bahia') >= '2022-06-01 00:00:00' and toStartOfDay(events.timestamp, 'America/Bahia') <= '2022-07-30 23:59:99' ) ) ) group by distinct_id, toStartOfDay(timestamp) ) group by time"`;

exports[`Analytics queries getConversionRateQuery should generate team conversion rate per month 1`] = `"select _moment as time, countIf(maxSteps >= 1) as renders, countIf(maxSteps = 2) as conversions, if (renders > 0, conversions / renders, 0) as conversion_rate from ( select distinct_id, toStartOfMonth(timestamp) as _moment, max(steps) as maxSteps from ( select distinct_id, timestamp, if(latest_render < latest_conversion, 2, if(latest_render is not null, 1, 0)) as steps from ( select distinct_id, timestamp, latest_render, min(latest_conversion) over ( partition by ( distinct_id, toStartOfMonth(timestamp) ) order by timestamp DESC ) latest_conversion from ( select distinct_id, timestamp, if(event = '$render', timestamp, null) as latest_render, if(event = '$conversion', timestamp, null) as latest_conversion from events where ( ( event = '$render' and has(JSONExtractArrayRaw(events.properties, 'teamIds'), '"gdYGJk5bkBPVEmFndhpoz4"') and JSONExtractString(events.properties, 'rootComponentId') = 'provider' ) or ( event = '$conversion' ) ) and toStartOfMonth(events.timestamp, 'America/Bahia') >= '2022-06-01' and toStartOfMonth(events.timestamp, 'America/Bahia') <= '2022-07-30' ) ) ) group by distinct_id, toStartOfMonth(timestamp) ) group by time"`;

exports[`Analytics queries getImpressionsQuery should generate project impressions per day sliced 1`] = `"select toStartOfDay(timestamp, 'America/Bahia') as time, JSONExtractString(events.properties, 'splitId') as slice, count(*) as impressions, count(distinct distinct_id) as unique_impressions from events where event = '$render' and has(JSONExtractArrayRaw(events.properties, 'teamIds'), '"gdYGJk5bkBPVEmFndhpoz4"') and has(JSONExtractArrayRaw(events.properties, 'projectIds'), '"kjuFFTSZb8fanzCHT2C1jz"') and JSONExtractString(events.properties, 'rootComponentId') = 'provider' and JSONHas(events.properties, 'splitId') and toStartOfDay(events.timestamp, 'America/Bahia') >= '2022-06-01 00:00:00' and toStartOfDay(events.timestamp, 'America/Bahia') <= '2022-07-30 23:59:99' group by time , slice"`;

exports[`Analytics queries getImpressionsQuery should generate project impressions per month sliced 1`] = `"select toStartOfMonth(timestamp, 'America/Bahia') as time, JSONExtractString(events.properties, 'splitId') as slice, count(*) as impressions, count(distinct distinct_id) as unique_impressions from events where event = '$render' and has(JSONExtractArrayRaw(events.properties, 'teamIds'), '"gdYGJk5bkBPVEmFndhpoz4"') and has(JSONExtractArrayRaw(events.properties, 'projectIds'), '"kjuFFTSZb8fanzCHT2C1jz"') and JSONExtractString(events.properties, 'rootComponentId') = 'provider' and JSONHas(events.properties, 'splitId') and toStartOfMonth(events.timestamp, 'America/Bahia') >= '2022-06-01' and toStartOfMonth(events.timestamp, 'America/Bahia') <= '2022-07-30' group by time , slice"`;

exports[`Analytics queries getImpressionsQuery should generate team impressions per day 1`] = `"select toStartOfDay(timestamp, 'America/Bahia') as time, count(*) as impressions, count(distinct distinct_id) as unique_impressions from events where event = '$render' and has(JSONExtractArrayRaw(events.properties, 'teamIds'), '"gdYGJk5bkBPVEmFndhpoz4"') and JSONExtractString(events.properties, 'rootComponentId') = 'provider' and toStartOfDay(events.timestamp, 'America/Bahia') >= '2022-06-01 00:00:00' and toStartOfDay(events.timestamp, 'America/Bahia') <= '2022-07-30 23:59:99' group by time"`;

exports[`Analytics queries getImpressionsQuery should generate team impressions per month 1`] = `"select toStartOfMonth(timestamp, 'America/Bahia') as time, count(*) as impressions, count(distinct distinct_id) as unique_impressions from events where event = '$render' and has(JSONExtractArrayRaw(events.properties, 'teamIds'), '"gdYGJk5bkBPVEmFndhpoz4"') and JSONExtractString(events.properties, 'rootComponentId') = 'provider' and toStartOfMonth(events.timestamp, 'America/Bahia') >= '2022-06-01' and toStartOfMonth(events.timestamp, 'America/Bahia') <= '2022-07-30' group by time"`;
